events/spamdetect.js

```diff
@@ -0,0 +1,85 @@
+// how many identical messages must be sent in a row for Wilson to take action
+const MESSAGES_TO_CACHE = 3
+// how long messages should be remembered at least (it will take between 1x and 2x this duration before they are removed from memory)
+const PURGE_INTERVAL_MINUTES = 3
+// this gets sent to the user whose messages have been detected as spam
+const MESSAGE_TO_USER = `As a spam prevention measure, you've been muted for sending the exact same message ${MESSAGES_TO_CACHE} times in a row. Please contact a moderator to be un-muted.`
+// messages in these channels are ignored
+const IGNORED_CHANNELS = [
+    '649186872026988557', // #botspam
+    '649186919317504012', // #random
+]
+// only check messages in these guilds
+const ACTIVE_GUILDS = [
+    '649165975647682560', // Wintergatan
+]
+const REPORT_CHANNEL = '709405505990426624' // #staff-botspam
+const MUTED_ROLE = '710120710084755477' // Muted
+
+
+const recently_posted = {
+    /* 'userid': [ recent messages, queue with length MESSAGES_TO_CACHE ] */
+}
+
+module.exports = async message => {
+	if (message.author.bot || IGNORED_CHANNELS.includes(message.channel.id)
+                           || !ACTIVE_GUILDS.includes(message.guild.id))
+		return;
+
+    // initialize message cache for user
+    if (!(message.author.id in recently_posted))
+        recently_posted[message.author.id] = []
+
+    // push message to queue
+    recently_posted[message.author.id].push(message)
+
+    // remove oldest item from queue if over max length
+    if (recently_posted[message.author.id].length > MESSAGES_TO_CACHE)
+        recently_posted[message.author.id].shift()
+
+    // check if all 3 messages are equal and are NOT stickers/attachments without message content
+    if (mightContainLink(message) && recently_posted[message.author.id].filter(stored => messagesEqual(message, stored)).length == MESSAGES_TO_CACHE) {
+        // delete the recent messages
+        for (stored of recently_posted[message.author.id]) {
+            stored.delete().catch(console.error)
+        }
+    
+        // Give member muted role
+        message.member.roles.add(MUTED_ROLE).catch(console.error)
+
+        // Report in #staff-botspam
+        message.guild.channels.resolve(REPORT_CHANNEL).send?.(`<@${message.author.id}> sent the same message ${MESSAGES_TO_CACHE}x in a row, in `
+        + `${recently_posted[message.author.id].map(m => m.channel.toString()).join(', ')}, content:\n\`\`\`\n${message.content}\n\`\`\``).catch(console.error)
+
+        // tell user they've been muted (try via dm, fall back to channel)
+        message.author.send(MESSAGE_TO_USER).catch(e => {
+            if (e.code == 50007)
+                message.reply(MESSAGE_TO_USER) 
+        })
+    }
+
+};
+
+function messagesEqual(a, b) {
+    // ignore stickers and attachment uploads without content
+    return a.content == b.content && a.content.length > 0
+}
+
+function mightContainLink(message) {
+    return message.content.includes('.')
```

Comment:
The get-urls package is an alternative, it has a list of TLDs which it checks -- should prevent false positives from any message with punctuation.

Reply:
I wanted to give `get-urls` a try to see how fast it is, but on [RunKit](https://npm.runkit.com/get-urls) it complains that it can't find `re2` and locally it gives the same message but with `node-re2`, and I can't seem to figure out how to fix it; `node-re2` isn't a package and installing `re2` doesn't help.

I think it won't be much of an issue if we don't do the url check, as sending 3 identical messages that do include a `.` (so just "yes" won't trigger it) in a row, and that _not_ being spam, isn't something that happens often (but maybe it has a legitimate purpose that I'm not thinking of).